### DISCUSSION TITLE
Raspberry Pi - Submissions for EXPORT competition

### DISCUSSION THREAD

**Discussion by batrlatom** [ultralytics/yolov5#3430] [Show and tell]
## <div align="center">Compete and Win</div>

Use this discussion thread for YOLOv5 🚀 EXPORT Competition submissions in the **Raspberry Pi** category. Good luck!

<p align="center">
  <a href="https://github.com/ultralytics/yolov5/discussions/3213">
  <img width="850" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/banner-export-competition.png"></a>
</p>

**Comment by karanjakhar** [comment]
https://github.com/karanjakhar/yolov5-export-to-raspberry-pi

Please have a look.

**Reply by batrlatom** [reply]
Thank you for your submission. At first glance it looks good!

**Reply by karanjakhar** [reply]
Thankyou :)

**Reply by nocolour** [reply]
Good. Later I can test it on other ARM64 board like rockchip ... (OrangePi & NanoPi)

**Reply by soumendukrg** [reply]
Is this export backward compatible with Armv7 on Raspberry Pi 3B+?

**Reply by karanjakhar** [reply]
@soumendukrg I haven't tested it. But it should work.

**Reply by rsd2007** [reply]
Hi @karanjakhar, is there a way I can test the model before deploying it on Raspberry Pi? Maybe on google colab or on my PC with Linux on it.

**Reply by karanjakhar** [reply]
@rajchem yes. In colab there is a command to run the converted model. Moreover, you can run it on a PC also.

**Reply by jaskiratsingh2000** [reply]
@karanjakhar If I have to test the mAP value on Raspberry Pi. How can I do that? I see with the Standard version it can easily be found as `python val.py --data DATA.yaml --weights WEIGHTS.pt` but how is it possible here? Also what is the value it is giving for you?

**Reply by dhunjoshi** [reply]
tf.lite.TFLiteConverter.from_keras_model(keras_model) NameError: name 'keras_model' is not defined

**Reply by jaskiratsingh2000** [reply]
@karanjakhar Please let me know. I am still looking forward to it

**Reply by jaskiratsingh2000** [reply]
> Hi @karanjakhar I have looked submissions including the Intel Vino one.
> All of them support only small model don't they?
> 
> Thanks for your contribution.

Hi, @aviezab I want to ask one thing that when I am trying to test the mAP value on Raspberry Pi it gives me 0 and if I check the mAP value on the server then it gives me the correct value which is 0.68 I am not able to understand that why it is behaving this way since the computer is a computer. So I would really appreciate it if you can provide me with this idea.

FYI, I am not a part of this competition I just work on experimenting within my research

**Reply by karanjakhar** [reply]
Hi @jaskiratsingh2000  we can check mAP. I haven't implemented that code in the repo. I will try to update it as soon as possible. I am occupied with work.

**Reply by karanjakhar** [reply]
> @karanjakhar Please let me know. I am still looking forward to it

sorry @jaskiratsingh2000 for the late response.

**Reply by jaskiratsingh2000** [reply]
Also @karanjakhar Would highly appreciate that if you can please help me out with an answer to this https://github.com/ultralytics/yolov5/discussions/3430#discussioncomment-1045705

**Reply by shivamswarnkar** [reply]
@karanjakhar did you benchmark your solution in raspi (or any other device)? If yes then what FPS did you get?

**Reply by rsd2007** [reply]
Hi @karanjakhar for inferencing on video, the code is loading the tflite model for each frame? That's why it is giving slow results as well.

**Reply by karanjakhar** [reply]
@rajchem  You are right. Thank you for pointing it out. I have updated it.

**Reply by rsd2007** [reply]
@karanjakhar can you help how can I adjust the code for INT8 quantized tflite model?

**Reply by jaskiratsingh2000** [reply]
@karanjakhar 
I just have a question that what is the core difference between the Pytorch Model and the Tflite model? As I could see that Pytorch model is not working good on the Raspberry Pi whereas the Tflite works better but I am not able to figure out at what are the reasons for that its working.

**Comment by AnhPC03** [comment]
Hello, I've create my repository to export YOLOv5 model from Pytorch framework to MNN framework (MNN is a lightweight deep neural network inference engine, embedded devices, such as IoT). Please take a look, feel free to comment or create an issue. Hope you enjoy it.
https://github.com/AnhPC03/yolov5-export-to-raspberry-pi-mnn

**Reply by batrlatom** [reply]
Thanks. We will look at it

**Reply by nocolour** [reply]
Good Job.
but slower than tflite. If can fine-tune it and increase the speed will be perfect.

**Reply by AnhPC03** [reply]
@nocolour I've not compared my model with tflite, I think you did. But my code default saves image/video/webcam into local. Maybe it makes my code slower than your tflite model? Or you had add --nodisplay --nosave when inference?

**Reply by shivamswarnkar** [reply]
@AnhPC03  good job man! btw, what FPS did you get (on the colab cpu and/or raspi)?

**Reply by jaskiratsingh2000** [reply]
Why there is a need to export Pytorch model to other for making it run on Raspberry Pi?

**Comment by jaskiratsingh2000** [comment]
Does anyone have any idea about this issue, please? - https://github.com/ultralytics/yolov5/issues/4167

Would highly appreciate it if someone can help in know this.

**Comment by jaskiratsingh2000** [comment]
@batrlatom @glenn-jocher I just have a key question that what are we trying to do with exportation of current Yolov5 into the exported version? What is the need to export for Raspberry Pi like devices?

What is the difference between current yolov5 and the exported version of yolov5?

Can anyone of you please let me know? I would highly appreciate that.

**Reply by batrlatom** [reply]
@jaskiratsingh2000 Hi, the goal is to make a yolov5 inference as fast as possible on Raspi. In normal situation, yolov5 is not quite realtime on such system, so applying quantization or any other technique which makes it faster is what it needs to be done.

**Reply by jaskiratsingh2000** [reply]
@batrlatom Can you be more specific please because I am a researcher working with Yolo and Raspberry Pi and this is something very important for me to understand in terms of more detailed specifications.
Because I am kinda stuck here - https://github.com/ultralytics/yolov5/issues/4167
Do you have any idea about this mentioned issue as well? That would really be of great help for me to understand.
Please let me know.

Thanks and Regards

**Reply by batrlatom** [reply]
@jaskiratsingh2000 I do not own raspi now, so I am unable to exactly try out the installation. But are you sure you have everything the same? The same checkout version, all libraries are on the same version? Computers are the same, but software layers could be different. If you do the ```pip list``` on both system, does it output the same version of all libraries? Did you try implementation of submissions from others? It looks like others were able to make it work. If you test the outputs visualy, are they really so different? The precission you mentioned would imply that raspi version does not output almost any rectangle correctly.

**Reply by jaskiratsingh2000** [reply]
@batrlatom Yes. Every Library on both the machines is the same. I tried the implementation of the others but none of them actually have implemented the Accuracy Performance and due to this reason only I was not able to understand the key question that what is the difference between the current Yolo version and the Exported version.
I would really appreciate it if you can help me with this issue since this is the core problem I see for Yolo on Rpi

**Reply by batrlatom** [reply]
in utils/general.py is a function `def non_max_suppression` . There is a variable time limit set to 10. Try to increase it, hopefully it helps with the mAP computation.

**Comment by jaskiratsingh2000** [comment]
@karanjakhar I am still looking forward to hearing your response on this - https://docs.google.com/document/d/1ej8hIktTFmsqTH0GCYVw-qD8sF9C9fa_Pe61EenNmkU/edit?usp=sharing

Can you please let me know?

If anyone else also has any idea about this then please let me know since I and @batrlatom have been discussing this but couldn't reach a conclusion.

Your response would be highly appreciable. Thanks!

**Reply by batrlatom** [reply]
@jaskiratsingh2000 Please solve the problem in the thread which is separate from the competition itself, thank you

**Comment by antran89** [comment]
Hi, I am a newbie, do we need to test it on Raspberry Pi device.

**Comment by blange48** [comment]
Hello, the new realease of yolo bring a amazing export model witch work perfectly on raspberry pi. With a 320 image, it is possible to reach 3 FPS and works perfectly with the raspberry camera.

**Reply by glenn-jocher** [reply]
@blange48 thanks for the feedback! What is your workflow for exporting models and running them on Pi?

**Reply by blange48** [reply]
I am simply using the TensorFlow exporter ;)
```
python3 models/tf.py --weights yolov5.pt --cfg models/yolov5s.yaml --tfl-int8  --ncalib 1000
```
To run the model on the raspberry pi I am using : 
```
python3 detect.py --weights yolov5s-int8.tflite --img 320  --device cpu
```
On the raspberry pi I am using the Qengineering distribution for raspberry pi. 
https://github.com/Qengineering


I am starting to test the same operation on the Jetson nano

**Reply by glenn-jocher** [reply]
@blange48 got it, thanks!

@zldrobit @batrlatom looks like tflite int8 is working well on Raspberry Pi, and the new automatic int8 determination also seems to be working :)

**Reply by jaskiratsingh2000** [reply]
@blange48 How did you figure out the mAP value? I am just a research intern and using this as a part of my studies. DO you have any idea of this or if you faced while testing? https://docs.google.com/document/d/1ej8hIktTFmsqTH0GCYVw-qD8sF9C9fa_Pe61EenNmkU/edit?usp=sharing

**Reply by blange48** [reply]
@jaskiratsingh2000 , hello sorry I did not try the val script and I do not have check the mAP value.

**Reply by jaskiratsingh2000** [reply]
@blange48 But do you have any idea of the problem I shared with you in the Google Doc? I need help that how can I rectify that.

**Reply by JNaranjo-Alcazar** [reply]
Just to clarify, is this the image you are using?
https://github.com/Qengineering/RPi-image

**Reply by blange48** [reply]
Hello, Yes I use the 64 bit image. 
maybe I have install some other library, but I need to check

**Reply by JNaranjo-Alcazar** [reply]
Okey! Thanks for the quick reply. I would appreciate if you could chek that. :)

**Reply by JNaranjo-Alcazar** [reply]
Hi again @blange48 , i have followed the steps you mentioned but I am strugglig with opencv-python installation.

When I run, detect.py, the script does not find that package and then it cannot be installed.

Any ideas in how to solve this problem?

Thanks

**Reply by jaskiratsingh2000** [reply]
Also, @batrlatom @blange48 @JNaranjo-Alcazar I am getting the following error when I am trying to run the above Export command which is mentioned above.

```
TFLite export failure: Multiple files match '../data/coco128.yaml', specify exact path: ['./yolov5s_saved_model/.│······························./data/coco128.yaml', './utils/../data/coco128.yaml', './data/../data/coco128.yaml', './models/../data/coco128.ya│······························ml', './runs/../data/coco128.yaml']                                                                              │······························Traceback (most recent call last):                                                                               │······························  File "models/tf.py", line 537, in <module>                                                                     │······························    with open(check_file(opt.source)) as f:                                                                      │······························  File "./utils/general.py", line 260, in check_file                                                             │······························    assert len(files) == 1, f"Multiple files match '{file}', specify exact path: {files}"  # assert unique       │······························AssertionError: Multiple files match '../data/coco128.yaml', specify exact path: ['./yolov5s_saved_model/../data/│······························coco128.yaml', './utils/../data/coco128.yaml', './data/../data/coco128.yaml', './models/../data/coco128.yaml', '.│······························/runs/../data/coco128.yaml']

```

**Reply by blange48** [reply]
Hello, 
I have checked some steps with this firmware. And I have install open cv.
Here is the list of the commands :

sudo apt-get install cmake gfortran libjpeg-dev libtiff-dev libgif-dev libavcodec-dev libavformat-dev libswscale-dev libgtk2.0-dev libcanberra-gtk* libxvidcore-dev libx264-dev libgtk-3-dev libtbb2 libtbb-dev libdc1394-22-dev libv4l-dev libopenblas-dev libatlas-base-dev libblas-dev liblapack-dev libhdf5-dev protobuf-compiler
wget -O opencv.zip https://github.com/opencv/opencv/archive/4.5.0.zip && wget -O opencv_contrib.zip https://github.com/opencv/opencv_contrib/archive/4.5.0.zip && unzip opencv.zip && unzip opencv_contrib.zip && mv opencv-4.5.0 opencv && \
wget -O opencv.zip https://github.com/opencv/opencv/archive/4.5.1.zip && wget -O opencv_contrib.zip https://github.com/opencv/opencv_contrib/archive/4.5.1.zip && unzip opencv.zip && unzip opencv_contrib.zip && mv opencv-4.5.1 opencv && mv opencv_contrib-4.5.1 opencv_contrib
free -m
cd opencv
mkdir build
cd build
cmake -D CMAKE_BUILD_TYPE=RELEASE         -D CMAKE_INSTALL_PREFIX=/usr/local         -D OPENCV_EXTRA_MODULES_PATH=~/opencv_contrib/modules         -D ENABLE_NEON=ON         -D WITH_OPENMP=ON         -D BUILD_TIFF=ON         -D WITH_FFMPEG=ON         -D WITH_GSTREAMER=ON         -D WITH_TBB=ON         -D BUILD_TBB=ON         -D BUILD_TESTS=OFF         -D WITH_EIGEN=OFF         -D WITH_V4L=ON         -D WITH_LIBV4L=ON         -D OPENCV_ENABLE_NONFREE=ON         -D INSTALL_C_EXAMPLES=OFF         -D INSTALL_PYTHON_EXAMPLES=OFF         -D BUILD_NEW_PYTHON_SUPPORT=ON         -D BUILD_opencv_python3=TRUE         -D OPENCV_GENERATE_PKGCONFIG=ON         -D BUILD_EXAMPLES=OFF ..
make -j4
sudo make install
sudo ldconfig
sudo apt-get update
sudo apt-get upgrade
sudo apt autoremove


I have also install TensorFlow 
sudo apt-get install gfortran
sudo apt-get install libhdf5-dev libc-ares-dev libeigen3-dev
sudo apt-get install libatlas-base-dev libopenblas-dev libblas-dev
sudo apt-get install liblapack-dev
sudo -H pip3 install --upgrade setuptools
sudo -H pip3 install pybind11
sudo -H pip3 install Cython==0.29.21
sudo -H pip3 install h5py==2.10.0
pip install gdown
sudo cp ~/.local/bin/gdown /usr/local/bin/gdown
gdown https://drive.google.com/uc?id=1WDG8Rbi0ph0sQ6TtD3ZGJdIN_WAnugLO
sudo -H pip3 install tensorflow-2.4.1-cp37-cp37m-linux_aarch64.whl

I have a working image, maybe I can share it ..

**Reply by JNaranjo-Alcazar** [reply]
That would be very nice @blange48 if you shared your running image. I would appreciate that 🥇

**Reply by JNaranjo-Alcazar** [reply]
Hi again! thanks @blange48 for the input, I managed to run a detection using the example you wrote. However, I get the following results in the zidane.jpg
![zidane](https://user-images.githubusercontent.com/62600201/133788628-41cee01d-bbd5-49b7-8947-d6ddbae0b094.jpg)

I do not get the class as a string, for instance, person...

Do you get the same results? Is this a tflite conversion limitation or should it work properly?

Thanks

**Reply by zldrobit** [reply]
@JNaranjo-Alcazar Class names are stored in PyTorch `*.pt` weight files. In the conversion to TensorFlow/TFLite, class names are stripped. You have to edit 
https://github.com/ultralytics/yolov5/blob/302a1b0bb03b3dbae0cb41f43b6c6c998070ff49/detect.py#L109 to get class names.

**Reply by JNaranjo-Alcazar** [reply]
Okey, that is what I was thinking.... Thanks for sharing! 👍🏼

**Reply by JNaranjo-Alcazar** [reply]
One question regarding execution times. The inference time using the demo images is around 450 ms using int8 tflite model.

Can someone give some advice in how to optimize this times and further steps?

Thanks a lot for the help of this community! 👍🏼

**Reply by blange48** [reply]
> That would be very nice @blange48 if you shared your running image. I would appreciate that 🥇

Hello, the image is available here : https://drive.google.com/file/d/1WrhaWJL2_eIJKJHgdf5KFnYqruFVlq2R/view?usp=sharing

**Reply by JNaranjo-Alcazar** [reply]
Nice! Thank you @blange48

**Reply by jaskiratsingh2000** [reply]
@blange48 I just have a quick question to understand regarding which I am having a hard time understanding it since long. What is actually inside the Pytorch model and Tflite model? I don't understand that what's the format inside the PyTorch model. When do we say that we are compiling the Pytorh model then what does happen intermediately?

Also, @blange48 Were you able to get the mAP value of any model on the RPi.

I am tagging other people if in case they have any idea about this then that would really lead to a healthy discussion. 
cc @zldrobit @JNaranjo-Alcazar @batrlatom @glenn-jocher @shivamswarnkar

**Comment by jaskiratsingh2000** [comment]
Hey Everyone! @glenn-jocher @zldrobit @blange48  I just have a question that what is the core difference between the Pytorch Model and the Tflite model? As I could see that Pytorch model is not working good on the Raspberry Pi whereas the Tflite works better but I am not able to figure out at what are the reasons for that its working.

**Comment by stefani-kovachevska** [comment]
Hi participants, 

Thank you all for your submissions, we have already reviewed them and are about to announce the winners! Please **send me an email** at [stefani@ultralytics.com](mailto:stefani@ultralytics.com) including your full name and GitHub username, so I can share more information with you. 

Stay tuned - the results are coming soon!

Stefani Kovachevska
_Project Manager at  Ultralytics_

**Comment by stefani-kovachevska** [comment]
Hi everyone, 

We were thrilled to have received your submissions to our YOLOv5 Export Competition for the Raspberry Pi category.

Unfortunately, none of the submissions passed the minimum set of requirements in each of the evaluation criteria. Therefore, no winner was selected for the category this time. However, there will be more chances for participants to compete again in the future. 

Thank you all for your efforts and for being a valuable member of our open-source community. Contributions from you are what makes our community great. 🙌 

Be sure to check out the winning repositories in the rest of the categories in our latest blog post: https://ultralytics.com/article/export-competition-winners

Stay amazing and keep creating! 🚀 
_Team Ultralytics_
